timemachine/fe/single_topology_v3.py

```diff
@@ -735,3 +742,182 @@ def setup_intermediate_state(self, lamb):
+        3) guest-guest interactions use pre-computed nonbonded interactions implemented as a pairlist.
+
+        Parameters
+        ----------
+        host_system: VacuumSystem
+            Parameterized system of the host
+
+        lamb: float
+            Which lambda value we want to generate the combined system.
+
+        Returns
+        -------
+        7-tuple of potentials
+            bond, angle, torsion, chiral_atom, chiral_bond, guest_nonbonded_precomputed, full_nonbonded
+
+        """
+
+        guest_system = self.setup_intermediate_state(lamb=lamb)
+
+        host_bond = host_system.bond
+        host_angle = host_system.angle
+        host_torsion = host_system.torsion
+        host_nonbonded = host_system.nonbonded
+
+        num_host_atoms = host_nonbonded.params.shape[0]
+
+        guest_bond = guest_system.bond
+        guest_angle = guest_system.angle
+        guest_torsion = guest_system.torsion
+        guest_nonbonded = guest_system.nonbonded
+        guest_chiral_atom = guest_system.chiral_atom
+        guest_chiral_bond = guest_system.chiral_bond
+
+        guest_chiral_atom_idxs = np.array(guest_chiral_atom.get_idxs(), dtype=np.int32) + num_host_atoms
+        guest_chiral_atom.set_idxs(guest_chiral_atom_idxs)
+
+        guest_chiral_bond_idxs = np.array(guest_chiral_bond.get_idxs(), dtype=np.int32) + num_host_atoms
+        guest_chiral_bond.set_idxs(guest_chiral_bond_idxs)
+
+        guest_nonbonded_idxs = np.array(guest_nonbonded.get_idxs(), dtype=np.int32) + num_host_atoms
+        guest_nonbonded.set_idxs(guest_nonbonded_idxs)
+
+        combined_bond_idxs = np.concatenate([host_bond.get_idxs(), guest_bond.get_idxs() + num_host_atoms])
+        combined_bond_params = np.concatenate([host_bond.params, guest_bond.params])
+        combined_bond = potentials.HarmonicBond(combined_bond_idxs).bind(combined_bond_params)
+
+        combined_angle_idxs = np.concatenate([host_angle.get_idxs(), guest_angle.get_idxs() + num_host_atoms])
+        combined_angle_params = np.concatenate([host_angle.params, guest_angle.params])
+        combined_angle = potentials.HarmonicAngle(combined_angle_idxs).bind(combined_angle_params)
+
+        if host_torsion:
+            combined_torsion_idxs = np.concatenate([host_torsion.get_idxs(), guest_torsion.get_idxs() + num_host_atoms])
+            combined_torsion_params = np.concatenate([host_torsion.params, guest_torsion.params])
+        else:
+            combined_torsion_idxs = np.array(guest_torsion.get_idxs(), dtype=np.int32) + num_host_atoms
+            combined_torsion_params = np.array(guest_torsion.params, dtype=np.float64)
+
+        combined_torsion = potentials.PeriodicTorsion(combined_torsion_idxs).bind(combined_torsion_params)
+
+        # concatenate ligand charges with host charges
+        ligand_exclusions = []
+        ligand_scale_factors = []
+
+        num_ligand_atoms = self.get_num_atoms()
```

Comment:
Style nit: "ligand" and "guest" both used this stretch to refer to the hybrid molecule.

(Possibly ambiguous collection of names: "guest", "ligand", "ligand_a", "ligand_b", "mol_a", "mol_b". Would there be any loss if we say "guest" throughout this function for hybrid molecule, "mol_a", "mol_b" for the endstates, and remove names "ligand", "ligand_a", "ligand_b"?)

Reply:
renamed to guest, I'm going to keep `mol_a` and `mol_b` for now. I actually didn't want to initially have the `SingleTopology` class have any host-guest semantics (and be focused only on pairs of small molecules).